### PR TITLE
tests: tallying and block finalization

### PR DESCRIPTION
**Pull request by gusin13**
This pr adds necessary assertions and code needed to make sure tallying and block finalization works

1. Inserts contract addresses in genesis (when bcd chain start, this is a hack ideally this would happen by gov proposal)
2. In the end check if the latest comet height is finalized (because tallying should work now)
3. Also changes the order of tests, this is in sync with this comment 
https://github.com/babylonchain/babylon-deployment/pull/99#issuecomment-2193909123

### CODE REVIEW

File: itest/cosmwasm/bcd/bcd_handler.go [hunk 1]
```diff
@@ -138,9 +139,37 @@ func bcdInit(homeDir string) error {
...
+		return fmt.Errorf("failed to update babylon_contract_address in genesis.json: %w", err)
+	}
+
+	// Update btc_staking_contract_address
+	sedCmd3 := fmt.Sprintf("sed -i. 's/\"btc_staking_contract_address\": \"\"/\"btc_staking_contract_address\": \"%s\"/' %s", btcStakingContractAddr, genesisPath)
+	fmt.Println("Executing command:", sedCmd3)
+	_, err = common.RunCommand("sh", "-c", sedCmd3)
+	if err != nil {
+		return fmt.Errorf("failed to update btc_staking_contract_address in genesis.json: %w", err)
+	}
```
Comment (SebastianElvis) [inline itest/cosmwasm/bcd/bcd_handler.go:170]: Great work on hacking this around :+1:

**Review by SebastianElvis — APPROVED** [review]
The code lgtm. Somehow the e2e test failed though. 

```
--- FAIL: TestConsumerFpLifecycle (0.02s)
    babylon_node_handler.go:145: 
                Error Trace:    /home/circleci/project/itest/babylon_node_handler.go:145
                                                        /home/circleci/project/itest/cosmwasm/bcd/bcd_test_manager.go:66
                                                        /home/circleci/project/itest/cosmwasm/bcd/bcd_consumer_e2e_test.go:32
                Error:          Received unexpected error:
                                exec: "babylond": executable file not found in $PATH
                Test:           TestConsumerFpLifecycle
FAIL
FAIL    github.com/babylonchain/finality-provider/itest/cosmwasm/bcd    0.226s
ok      github.com/babylonchain/finality-provider/keyring       0.964s
FAIL
```

maybe somehow `babylond` is not installed?

**Comment by maurolacy** [comment]
Tests are broken atm. Contracts seems to be deployed into the wrong chain (Babylon chain instead of `bcd` Consumer chain).

**Comment by gusin13** [comment]
> Tests are broken atm. Contracts seems to be deployed into the wrong chain (Babylon chain instead of bcd Consumer chain).

contract is deployed on bcd [ref](https://github.com/babylonchain/finality-provider/pull/446/files#diff-907f50bcb7070a373182804218ceda1af8fa5f7a9899e2f2012b36b516644b10R48), the problem seems to be something else. The bcd cmd package is not being installed properly in ci jobs, if I run the tests locally they are passing. On my local I manually install the bcd by cloning babylon-sdk and then make install, but if i use `go install github.com/babylonchain/babylon-sdk/demo/cmd/bcd` then it errors out not sure why this is happening. 

If the package is install using go install (like its done in ci jobs), then 
```
gusin@Gurjots-MacBook-Pro bin % bcd version

```

if its done manually using make install (by cloning babylon-sdk and cd babylon-sdk then make install)
```
gusin@Gurjots-MacBook-Pro bin % bcd version
v0.2.0-rc.1
```

trying to debug why this is happening

File: itest/cosmwasm/bcd/bcd_handler.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ func NewBcdNodeHandler(t *testing.T) *BcdNodeHandler {
 
 	setupBcd(t, testDir)
```
[the remark this answers is not in the document]
Comment (gusin13) [reply itest/cosmwasm/bcd/bcd_handler.go:38]: > This requires that bcd is built and installed


bcd command is installed using make target then later the node is started using bcd cli

> Let's use a submodule for it, along with a commit / tag reference, and add the build instructions to the test and test-e2e targets.

not sure if i got this properly, can you elaborate how would a submodule be used here? Docker image is a good idea similar to the way tests are done in babylon repo, but this requires major refactoring

**Comment by bap2pecs** [comment]
@gusin13 op tests failed likely due to rate limiting querying BTC mainnet RPCs. we fixed it already to mock the RPC client. after you rebase, it should be fixed

File: itest/cosmwasm/bcd/bcd_handler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gusin13) [reply itest/cosmwasm/bcd/bcd_handler.go:43]: have opened a issue to track this probably will be done in a separate pr
https://github.com/babylonchain/finality-provider/issues/477

File: itest/bytecode/version.txt
[the remark this answers is not in the document]
Comment (gusin13) [reply itest/bytecode/version.txt:1]: i intend to use a newer commit than the tagged one, maybe once we have the next release in babylon-contract i'll bump the version here as well

**Pull request merged**